lightpath/controller.py

```diff
@@ -42,115 +49,402 @@ def __init__(self, client, endstations=None):
-                         line, start, end)
-            results = self.client.search_range(key='z', start=start, end=end,
-                                               beamline=line, active=True,
-                                               lightpath=True)
-            # Ensure we actually found valid devices
-            if not results:
-                logger.error("No valid beamline devices found for %s", line)
-                continue
-            # Load all the devices we found
-            logger.debug("Found %s devices along %s", len(results), line)
-            for result in results:
-                try:
-                    dev = result.get()
-                    devices.append(dev)
-                except Exception:
-                    logger.exception("Failure loading %s ...", result["name"])
-                    self.containers.append(result.device)
-        # Create the beamline from the loaded devices
-        bp = BeamPath(*devices, name=line)
-        self.beamlines[line] = bp
-        # Set as attribute for easy access
-        setattr(self, bp.name.replace(' ', '_').lower(), bp)
-        return bp
+                           "assuming this is an invalid path", endstation)
+            return
 
-    @property
-    def destinations(self):
+        paths = list()
+        for branch in end_branches:
+            # Find the paths from each source to the desired line
+            for src in self.sources:
+                if nx.has_path(self.graph, src, branch):
+                    found_paths = nx.all_simple_paths(self.graph,
+                                                      source=src,
+                                                      target=branch)
+                    paths.extend(found_paths)
+                else:
+                    logger.debug(f'No path between {src} and {branch}')
+
+        self.beamlines[endstation] = []
+        for path in paths:
+            subgraph = self.graph.subgraph(path)
+            devices = [node[1]['dev'] for node in subgraph.nodes.data()
+                       if node[1]['dev'] is not None]
+            bp = BeamPath(*devices, name=endstation)
+            self.beamlines[endstation].append(bp)
+
+        # This isn't used currently, but left for now
+        setattr(self, bp.name.replace(' ', '_').lower(),
+                self.beamlines[endstation])
+
+    def imped_z(self, path: BeamPath) -> float:
         """
-        Current device destinations for the LCLS photon beam
+        Get z position of impediment or inf.
+
+        Parameters
+        ----------
+        path : :class:`BeamPath`
+            BeamPath to find impediment position in
+
+        Returns
+        -------
+        float
+            z position of impediment
+        """
+        return getattr(path.impediment, 'md.z', math.inf)
+
+    def active_path(self, dest: str) -> BeamPath:
+        """
+        Return the most active path to the requested endstation
```

Comment:
Is this intended to help for cases like XCS where there are 2 routes?

Reply:
That was the idea.  The tests for this method primarily use XCS routes.  (and they pass I promise)